### PR TITLE
[MX-9588] Add micro averaging strategy for F1 metric

### PR DESCRIPTION
**Pull request by sethah**
## Description ##
This PR adds a mixin class that F1 and other metrics like precision and recall can leverage in the future. It also provides a new option for the F1 metric called `average` which defines how the metric will be aggregated across mini batches. 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Approach

The "micro" vs "macro" update strategy is not specific to F1 score. The macro update just takes an average of averages, which can be done for any metric. It may be best to design an abstraction where any metric can have the micro/macro update option, but I couldn't see a good way to do that here that would:

* be easy to use for end users AND
* maintain backward compatibility AND
* maintain current semantics

For now, the behavior for each type of update is hard coded into the `update` method of the `F1` class. We can discuss the approach.

Please let me know if I have missed or overlooked anything :)

### CODE REVIEW

**Comment by sethah** [comment]
Regarding other approaches, something I looked at was the following:

```python
class MacroMetric(EvalMetric):

    def __init__(self, base_metric):
        super(MacroMetric, self).__init__("macro_" + base_metric.name, output_names=base_metric.output_names,
                                          label_names=base_metric.label_names)
        self.base_metric = base_metric

    def update(self, labels, preds):
        self.base_metric.update(labels, preds)
        self.sum_metric += self.base_metric.get()[1]
        self.num_inst += 1
        self.base_metric.reset()
```

Any metric that has defined the "micro" behavior can then be used as "macro" just by calling `metric = mx.metric.MacroMetric(mx.metric.F1())`, but that seems pretty awkward for the users, and also changes the default behavior to micro, which doesn't work for backwards compatibility. The current solution works well enough, and we could probably revisit in a later PR if needed.

**Comment by szha** [comment]
This fixes #9588

File: python/mxnet/metric.py [hunk 1]
```diff
@@ -475,8 +475,84 @@ def update(self, labels, preds):
...
+        pred = pred.asnumpy()
+        label = label.asnumpy().astype('int32')
+        pred_label = numpy.argmax(pred, axis=1)
+
+        check_label_shapes(label, pred)
+        if len(numpy.unique(label)) > 2:
+            raise ValueError("%s currently only supports binary classification."
+                             % self.__class__.__name__)
+
+        for y_pred, y_true in zip(pred_label, label):
```
Comment (sxjscience) [inline python/mxnet/metric.py]: Do we have to use a for-loop here? Using array arithmetic OPs should be more efficient. Also we can try not to convert them into numpy array so it can be calculated in GPU.

Comment (sethah) [reply python/mxnet/metric.py]: I agree, but there's another issue for that, so I assumed it would be done in a separate PR. https://github.com/apache/incubator-mxnet/issues/9586

File: tests/python/unittest/test_metric.py [hunk 2]
```diff
@@ -56,18 +55,51 @@ def test_acc():
...
+    np.testing.assert_almost_equal(macroF1.get()[1], fscore1)
+
+    microF1.update([label21, label22], [pred21, pred22])
+    macroF1.update([label21, label22], [pred21, pred22])
+    assert microF1.num_inst == 6
+    assert macroF1.num_inst == 2
+    fscore2 = 2. * (1) / (2 * 1 + 0 + 0)
+    fscore_total = 2. * (1 + 1) / (2 * (1 + 1) + (1 + 0) + (0 + 0))
+    np.testing.assert_almost_equal(microF1.get()[1], fscore_total)
+    np.testing.assert_almost_equal(macroF1.get()[1], (fscore1 + fscore2) / 2.)
```
Comment (sxjscience) [inline tests/python/unittest/test_metric.py]: For the test part I think one way is to compare the result with [sklearn.metrics.f1_score](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html). However I'm not sure if it works for CI. @marcoabreu

File: python/mxnet/metric.py (hunk 1 above)
Comment (szha) [reply python/mxnet/metric.py]: Let's address that in a separate PR. My last attempt faced performance issues when switching to ndarray-based logic, so we should address the problem once it's more clear how to resolve the performance issue.

**Comment by sxjscience** [comment]
I think the current way is fine. All we need to do is to provide a mx.metric version of sklearn.metrics.f1_score.

File: python/mxnet/metric.py [hunk 3]
```diff
@@ -475,8 +475,84 @@ def update(self, labels, preds):
             self.num_inst += num_samples
 
 
+class _BinaryClassificationMixin(object):
+    """
+    Private mixin for keeping track of TPR, FPR, TNR, FNR counts for a classification metric.
```
Comment (piiswrong) [inline python/mxnet/metric.py]: Could you add more explanation of what this does?

File: python/mxnet/metric.py [hunk 4]
```diff
@@ -475,8 +475,84 @@ def update(self, labels, preds):
...
+
+    def _reset_stats(self):
+        self._false_positives = 0
+        self._false_negatives = 0
+        self._true_positives = 0
+        self._true_negatives = 0
+
 @register
-class F1(EvalMetric):
+class F1(EvalMetric, _BinaryClassificationMixin):
```
Comment (piiswrong) [inline python/mxnet/metric.py]: use member instead of multiple inheritance if possible

Comment (szha) [reply python/mxnet/metric.py]: This is one of the cases where usage of mix-ins is proper for backward compatibility and ease of use.

Comment (sethah) [reply python/mxnet/metric.py]: I think we can make a `BinaryClassificationMetric` class that is intended to be abstract and inherits from `EvalMetric` without affecting backward compatibility. It seemed more appropriate as a mixin here since it would be useless as a concrete class and doesn't actually implement any functionality required by `EvalMetric`.

File: python/mxnet/metric.py (hunk 3 above)
Comment (sethah) [reply python/mxnet/metric.py]: Updated. Let me know if that's not what you had in mind.

File: python/mxnet/metric.py (hunk 1 above)
Comment (sxjscience) [reply python/mxnet/metric.py]: I've used codes like 
```python
tp = nd.sum((pred == 1) * (label == 1)).asscalar()
fp = nd.sum((pred == 1) * (label == 0)).asscalar()
fn = nd.sum((pred == 0) * (label == 1)).asscalar()
precision = float(tp) / (tp + fp)
recall = float(tp) / (tp + fn)
f1 = 2 * (precision * recall) / (precision + recall)
```
to calculate the F1 and I find it's much faster in GPU.

**Comment by sxjscience** [comment]
Also, one thing to note is that when the output has only one-label, the micro F1 is equivalent to accuracy https://stackoverflow.com/questions/37358496/is-f1-micro-the-same-as-accuracy .

File: python/mxnet/metric.py (hunk 1 above)
Comment (sethah) [reply python/mxnet/metric.py]: Shouldn't this conversation take place on #9586? This PR only aims to fix #9588.

Comment (sxjscience) [reply python/mxnet/metric.py]: OK, it should be addressed in a later PR.

**Comment by szha** [comment]
@sxjscience let's track the multi-label case in #9589

**Comment by szha** [comment]
@sethah do you mind running the related test from #9705 and report the before/after change results here?

File: tests/python/unittest/test_metric.py (hunk 2 above)
Comment (marcoabreu) [reply tests/python/unittest/test_metric.py]: Since we're lacking dependency support on Windows slaves yet, we can't use this if scikit is not present as a dependency yet - I can't check right now. I'd propose to just try it out and see whether it works or not.

**Comment by sethah** [comment]
Obviously, no change is expected since none of the update logic was changed here.

**Before**
```
Metric         Data-Ctx  Label-Ctx   Data Size   Batch Size     Output Dim     Elapsed Time
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      131072      16             2              0.82034
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      131072      64             2              0.46931
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      131072      256            2              0.35179
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      131072      1024           2              0.35274
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      16384       16             2              0.10991
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      16384       64             2              0.057258
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      16384       256            2              0.046497
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      16384       1024           2              0.044378
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      2048        16             2              0.01292
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      2048        64             2              0.0070581
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      2048        256            2              0.005384
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      2048        1024           2              0.00511
------------------------------------------------------------------------------------------
```

**After**
```
Metric         Data-Ctx  Label-Ctx   Data Size   Batch Size     Output Dim     Elapsed Time
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      131072      16             2              0.81942
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      131072      64             2              0.46574
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      131072      256            2              0.37653
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      131072      1024           2              0.32401
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      16384       16             2              0.099127
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      16384       64             2              0.057564
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      16384       256            2              0.044527
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      16384       1024           2              0.04604
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      2048        16             2              0.012832
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      2048        64             2              0.0071719
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      2048        256            2              0.0058281
------------------------------------------------------------------------------------------
F1             cpu(0)    cpu(0)      2048        1024           2              0.0057449
------------------------------------------------------------------------------------------
```

File: tests/python/unittest/test_metric.py (hunk 2 above)
Comment (sethah) [reply tests/python/unittest/test_metric.py]: Ok, trying this out. We'll see.

Comment (sxjscience) [reply tests/python/unittest/test_metric.py]: Seems that sklearn is not installed in some machines.

Comment (sethah) [reply tests/python/unittest/test_metric.py]: Yeah, I've just reverted the commit.

File: python/mxnet/metric.py (hunk 4 above)
Comment (piiswrong) [reply python/mxnet/metric.py]: multiple inheritance is very rarely necessary.
In this case I think _BinaryClassificationMixin should either inherit EvalMetric or be refactored into a few utility functions

Comment (piiswrong) [reply python/mxnet/metric.py]: composition is also possible

Comment (szha) [reply python/mxnet/metric.py]: Honestly, the current multiple inheritance seems reasonable, in that calculating counts and keeping counts are two separate concerns. Mixins is more flexible and likely require less code when we extend these to multi-class/multi-label/top-k use cases.

Comment (szha) [reply python/mxnet/metric.py]: Composition is equally flexible with the drawback of extra dereference, though I'm fine either way.

Comment (sethah) [reply python/mxnet/metric.py]: Can you check latest commit to see if that's what you had in mind?

File: python/mxnet/metric.py [hunk 5]
```diff
@@ -503,21 +582,27 @@ class F1(EvalMetric):
     label_names : list of str, or None
         Name of labels that should be used when updating with update_dict.
         By default include all labels.
+    average : str, default 'macro'
+        Strategy to be used for aggregating across micro-batches.
```
Comment (szha) [inline python/mxnet/metric.py]: "mini-batches" is more commonly used.

File: python/mxnet/metric.py [hunk 6]
```diff
@@ -503,21 +582,27 @@ class F1(EvalMetric):
...
+            "macro": average the F1 scores for each batch
+            "micro": compute a single F1 score across all batches
```
Comment (szha) [inline python/mxnet/metric.py]: Add period at the end. Currently it renders into:
http://mxnet-doc.s3-accelerate.dualstack.amazonaws.com/api/python/metric/metric.html#mxnet.metric.F1

**Comment by sethah** [comment]
Thanks all for the review!
